### PR TITLE
DATAINT-1747: Add Dynamic Date Query Support

### PR DESCRIPTION
**Pull request by roehlerw**
resolves DATAINT-1747

Added support for dynamic dates in custom queries

### CODE REVIEW

File: PluginAutotask/API/Utility/ApplyDynamicDate.cs [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+                return query;
+            }
+
+            foreach (var filter in query.Filter)
+            {
+                if (filter.Value is string)
+                {
+                    var value = (string)filter.Value;
+                    var parts = value.Split('_');
+                    if (parts.Length != 3)
```
Comment (g1hanna) [inline PluginAutotask/API/Utility/ApplyDynamicDate.cs]: I do not see a check for the final element of the split string (`parts[2]`). As part of the unit test, a user could type in something other than DAYS as the final part of the constant (e.g., TODAYMINUS_N_**MINUTES**) and still have the unit test pass.

Edit: Please let me know if this isn't a concern and I'll update my review.

**Review by g1hanna — CHANGES_REQUESTED** [review]

File: PluginAutotask/API/Utility/ApplyDynamicDate.cs (hunk 1 above)
Comment (roehlerw) [reply PluginAutotask/API/Utility/ApplyDynamicDate.cs]: Good point. I have added more conditions to fall into the error note.

**Review by g1hanna — APPROVED** [review]

File: PluginAutotask/API/Utility/ApplyDynamicDate.cs
Comment (uttam028) [inline PluginAutotask/API/Utility/ApplyDynamicDate.cs:1]: An alternate implementation:

//You can restrict the max days up to 999 or change them accordingly
public static string pattern = @"^TODAYMINUS_[0-9]{1,3}_DAYS$";
....
....
// remove RegexOptions.IgnoreCase if it must be capital
if (parts.Length != 3 || !int.TryParse(parts[1], out var numDays) || !Regex.IsMatch(value, pattern, RegexOptions.IgnoreCase))
{
            throw new Exception......
}

var dynamicDate = DateTime.Today.AddDays(-numDays);

**Review by uttam028 — APPROVED** [review]

File: PluginAutotask/API/Utility/ApplyDynamicDate.cs
Comment (uttam028) [reply PluginAutotask/API/Utility/ApplyDynamicDate.cs:1]: The current code will get an integer parse exception if the filter value is "TODAYMINUS_abc_DAYS"

Comment (roehlerw) [reply PluginAutotask/API/Utility/ApplyDynamicDate.cs:1]: I think this is a great addition and has been added.

**Review by g1hanna — APPROVED** [review]
Lovely addition with the Regex rule

**Pull request merged**